Feature/14 review shiny ci cd (#27)
miraisolutions/techguides: shiny-ci-cd.Rmd

> Most of these steps are implemented by default in Travis CI for an R package. In GitHub Actions, on the other hand, it is currently necessary to manually specify each of them.

Author: I disagree, the sentence above introduces the list of steps and then this sentence directly below it refers to those steps, so it should be clear what is meant and the order makes sense to me as well (i.e. mentioning how these steps relate to github actions and travis CI as a further, noteworthy detail).

## GitHub Actions

[GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions) is a service for running highly-customizable and flexible automated workflows, fully integrated with GitHub and very suitable to CI/CD pipelines.
[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.

> Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
> 

Reviewer: ```suggestion
Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.

The next sections describe in detail the relevant workflow steps of a typical CI/CD pipeline for a packages Shiny app, also covering the usage of `renv` to track package dependencies. Finally, we will show how you can use the convenience function `usethis::use_github_action()` for including such workflows in you project.
```

### Workflow steps

> A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.

Reviewer 2: Worth to mention somewhere that Actions in the workflow can be pulled from somewhere else and combined.

> A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.

Author: Not sure what you mean by that. Can you either add some text yourself for this or elaborate / link to somewhere where this is explained?

> A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.

Reviewer 2: I meant that one can take an action defined by someone and use it in his/her own workflow.

> A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be a (predefined) action or a script, however, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being fully customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.

Author: I see. I guess it is somewhat implied, but surely doesn't harm to point that out. I updated this with another commit.

> A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be an action (predefined, sourcing from GitHub repos that contain such actions) or a script. However, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being entirely customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.

Reviewer: There is something going on by now, although not mature yet
```suggestion
A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be an action (predefined, sourcing from GitHub repos that contain such actions) or a script. However, for the time being, GitHub Actions does not provide a mature support for hierarchical, aggregated structure of actions. Being entirely customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
```

#### Setup

> - Checkout the source package from the repo, using `actions/checkout` provided by GitHub.

Reviewer 2: ```suggestion
- Checkout the source package from the current repo, using the `actions/checkout` provided by GitHub.
```

Reviewer 2: If this step also includes ubuntu installation, we should mention it explicitly.

Author: It does not. I don't like the added article, will remove it on a later commit.

> - Checkout the source package from the current repo, using `actions/checkout` provided by GitHub.

Reviewer: We are talking about "the repository" in the intro, there is no "current repository" for a techguides reader 
```suggestion
- Checkout the source package from the repository, using `actions/checkout` provided by GitHub.
```

Author: we are not consistent with repository / repo in all our articles (perhaps not in the techguides, just remember seeing both being used). I would favor repository for our techguides and articles.

> - Setup R using action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).

Reviewer 2: ```suggestion
- Setup R using the action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).
```

Author: Here adding the article is fine, since there is also the noun "action" before the actual "name" / grammatical thing.

- Query and cache R package dependencies using `remotes::dev_package_deps()` and the `actions/cache` predefined action.

> - Install system dependencies using package `sysreqs::sysreq_commands()` (for the `ubuntu` runner used to run the action).

Reviewer: ```suggestion
- Install system dependencies (for the `ubuntu` runner defined for the workflow) using `sysreqs::sysreq_commands()`.
```

> - Install R package dependencies using `remotes::Install_deps()`.

Reviewer: ```suggestion
- Install R package dependencies using `remotes::install_deps()`.
```

#### Package check

> - Check the package via using `rcmdcheck::rcmdcheck()`.

Reviewer 2: ```suggestion
- Check the package via `rcmdcheck::rcmdcheck()`.
```

#### Deployment

> - Deploy to shinyapps.io, similar to the Travis CI approach:

Reviewer 2: ```suggestion
- Deploy to shinyapps.io:
```
Reference to TravisCI is not needed anymore (also believe it is a legacy from when Travis was the default case upon which GitHub actions was building up)

>   - In this case, environment variables `SHINYAPPS_ACCOUNT`, `SHINYAPPS_TOKEN` and `SHINYAPPS_SECRET` (defining credentials for [shinyapps.io](https://www.shinyapps.io/)), are specified / accessible as GitHub [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference#authentication-and-secrets).

Reviewer: We need to introduce this better, as was originally done with the Travis as the first part of the guide
```suggestion
  - In order to provide credentials for the deployment, account name and corresponding [tokens](https://docs.rstudio.com/shinyapps.io/getting-started.html#deploying-applications) for shinyapps.io, environment variables `SHINYAPPS_ACCOUNT`, `SHINYAPPS_TOKEN` and `SHINYAPPS_SECRET` are specified / accessible as GitHub [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference#authentication-and-secrets).
```

  - An R script e.g. `deploy/deploy-shinyapps.R` (build-ignored via `usethis::use_build_ignore("deploy")`) defines the deployment commands based on the environment variables:
```{r read-deploy-script, eval = TRUE, echo = FALSE}
knitr::read_chunk("shiny-ci-cd/deploy/deploy-shinyapps.R", labels = "deploy-shinyapps")
```
```{r deploy-shinyapps}
```

### Workflow file

> The `steps` above are defined in the `.yml` workflow file as follows:

Reviewer 2: ```suggestion
The `steps` described in the previous section are defined in the `.yml` workflow file as follows:
```

Reviewer 2: not sure how to reference the section directly

```{r read-ci-cd.yml, eval = TRUE, include = FALSE}
file.copy("shiny-ci-cd", "_book", recursive = TRUE)
knitr::read_chunk("shiny-ci-cd/actions/ci-cd.yml", labels = "ci-cd.yml")
```
```{yml ci-cd.yml}
```

> As visible from the run logs, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field.

Reviewer 2: You have never mentioned the run logs before, so if you want to mention them you should explain them I believe.

Author: I updated this and it is more explicit now.

> As visible from the run logs that can be found on GitHub under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:

Reviewer: ```suggestion
As visible from the run logs that can be found in the GitHub repository under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:
```

> ![GitHub Actions Continuous Integration / Continuous Deployment pipeline for a packaged Shiny app](shiny-ci-cd/img/ShinyCICD_githubactions1.png)

Reviewer: We need a new image now to be aligned with the steps as revised

Author: I could take a screenshot from https://github.com/miraisolutions/ShinyCICD/runs/2508242565. For some reason, the title changed from CI-CD / CI-CD to ubuntu-20.04 (release). Let me know if we should investigate this or just assume that something changed in the background on how this title is derived and that it is fine to keep the new name.

Reviewer: I agree that a screenshot from the run you propose will do.

> the title changed from CI-CD / CI-CD to ubuntu-20.04 (release)

This is due to the fact that we set the `job`  `name: ` to the matrix entries for OS and R version. This makes sense and I think it is OK. CI-CD is also the name of the workflow and is now shown in the left side bar and not in the title of the run  output, which now only contains the job name.

Author: updated and pushed

> 

Reviewer: We completely miss the part about `renv`, which is very relevant to us. We can introduce this new section in the very same way as currently done in the Travis part (perhaps making that part slimmer do avoid redundancy), and highlight which steps are different (caching and restoring locked dependencies). Not sure we want to show the full worfklow, maybe just the `ci-renv.yml`, since deployment is unaffected

Author: This requires some thinking, as including the full ci-cd-renv.yml file just after ci-cd.yml seems a bit excessive (same with including a new screenshot for the renv steps). We could use collapsible sections for this or even put it there entirely, not sure what looks more appealing.

Reviewer: I suggest to go ahead with the introduction, just mentionthat `renv` requires a different caching and that dependencies are installed via `renv::restore`, pointing to https://rstudio.github.io/renv/articles/ci.html#github-actions.

I would not include new screenshots here, and would avoid the the full workflow in the `renv` section. Still, I think it makes sense to perhaps enhance the section about `usethis` to give more relevance to the 4 full workflow .yml files we provide and show them, indeed using collapsible / accordion. We could do this in R markdown using bsplus (http://ijlyttle.github.io/bsplus/index.html), see how tit was done for the CovidR contest in  https://github.com/Milano-R/erum2020-covidr-contest/blob/a13b0ee2b2eec57bc523c7d6e50734863e35a9d8/contest/index.Rmd#L86-L103

Overall we could have stg like 

```md

### Complete workflows and `usethis::use_github_action()`

Full YAML workflows for CI and CI/CD pipelines, with and without `renv` are shown below and provided as part of this guide.

In order to setup and use CI/CD GitHub Actions workflows as described above, you can simply include the relevant workflow file your project via:

[...] existing code snippet

[...] collapsible panel for each workflow files (perhaps within a dedicated `####` sub section), the panel (and subsection) title being the file name and the content is the file content

Author: @riccardoporreca I updated part of this, but could not add collapsible sections using bsplus (nor other approaches I tried from some quick googling + trying out) - it seems that collapsible sections don't work with bookdown and / or gitbook.

Before I do a more thorough investigation, would you know something about this? Maybe some work-around with html and possibly other things? I was searching for something about "collapsible sections" in the bookdown documentation, but could not find anything helpful regarding this problem.

Reviewer: - Collapsible full workflow files included using the `<details>` (https://www.w3schools.com/tags/tag_details.asp) in commit 19b455d 
- Explicit section about renv under GitHub action included as hinted above in commit 900444e

### `usethis::use_github_action()`

> In order to use the GitHub action workflow above, or its renv-based variant, you can simply:

Reviewer: ```suggestion
In order to setup and use CI/CD GitHub Actions workflows as described above, we provide full YAML workflow files for CI and CI/CD pipelines with and without `renv`, which you can simply include in your project via:
```

```{r}
usethis::use_github_action(
  url = 
    "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci-cd.yml"
  # "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci-cd-renv.yml"
  # "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci.yml"
  # "https://mirai-solutions.ch/techguides/shiny-ci-cd/actions/ci-renv.yml"
)
usethis::use_github_actions_badge("CI-CD") # or "CI"
```

## Travis CI

> Travis CI is a previously fully open-source continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).

Reviewer 2: ```suggestion
Travis CI is a continuous integration service, which used to be fully open-source, that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).
```

> Travis CI is a continuous integration service, which used to be fully open-source, that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).

Reviewer: The focus in not really on Travis no being open-source, but not really targeting / valuable for open-source projects.
```suggestion
Travis CI is a continuous integration service that can be used to build and test software projects hosted on GitHub. To set up Travis CI you need to login at [https://travis-ci.com/](https://travis-ci.com/) (using your GitHub account) and provide authorization via GitHub (see [Travis CI Tutorial](https://docs.travis-ci.com/user/tutorial)).

Travis CI used to be a very established, mature and popular tool in the open-source source community, before a recent change of policy made it less focused on open-source, offering only limited free trial plans.
```

>     branch: master

Reviewer 2: maybe somewhere we should add a note about the master branch being now called main if you create a new repo on GitHub

Author: Resolved as belongs to same topic of adding more comments to these files in general.
